log-elasticsearch/src/Log/Backend/ElasticSearch/V5.hs

```diff
@@ -144,19 +143,29 @@ elasticSearchLogger esConf@ElasticSearchConfig{..} genRandomWord = do
           $ "ElasticSearch: unexpected response: " <> jsonToBSL replyBody
         Just (hasErrors, items) -> when hasErrors $ do
           -- If any message failed to be inserted because of type mismatch, go
-          -- back to them, replace their data with elastic search error and put
-          -- old data into its own namespace to work around insertion errors.
+          -- back to them, log the insertion failure and add type suffix to each
+          -- of the keys in their "data" fields to work around type errors.
           let failed = V.findIndices (H.member "error") items
-          dummyMsgs <- V.forM failed $ \n -> do
-            dataNamespace <- liftIO genRandomWord
-            let modifyData oldData = object [
-                    "__es_error" .= H.lookup "error" (items V.! n)
-                  , "__es_modified" .= True
-                  , ("__data_" <> showt dataNamespace) .= oldData
-                  ]
-            return . second (H.adjust modifyData "data") $ jsonMsgs V.! n
+              newMsgs = (`V.map` failed) $ \n ->
+                let modifyData :: Value -> Value
+                    modifyData (Object hm) = Object $ H.fromList
+                      [ "__es_error" .= H.lookup "error" (items V.! n)
+                      , "__es_modified" .= True
+                      ] `H.union` H.foldlWithKey' keyAddValueTypeSuffix H.empty hm
+                    modifyData v = v
+
+                    keyAddValueTypeSuffix acc k v = H.insert
```

Comment:
I guess now we create a possibility for a corner case that would fail with that mechanism. Imagine we have a field `a` which is a `String`. Then we want to erroneously insert field `a` which is an `Object` and it's nested. So we insert `a_object.b` and `a_object.c`, where `b` and `c` are `String` for example. Next time we change the type of `b` from `String` to `Bool` and now we get another insertion error, we try `a_object.b` and it fails, because theoretically it should be `a_object.b_bool`. I hope it shouldn't happen, because the field `a` should be renamed asap after the error. But still, it's a possible corner case. I don't think we need to take it into consideration though.

Reply:
Good catch, I forgot about recursive case. Fixed in additional commit.